### PR TITLE
[NNX] Checkpoint restore: error on weights missing from the checkpoint

### PR DESCRIPTION
**Pull request by ecnal-cienet**
## Description

On the pure_nnx resume path, a weight present in the model but absent from the checkpoint is handled unsafely. Now that proactive `scan_layers` validation has landed (#4304) and the reactive mismatch guards were removed (#4340), `_populate_pure_dict_from_partial` fills every absent leaf with a default:
- a weight Orbax omits entirely → silently zero-filled (silent accuracy loss);
- a weight Orbax leaves under `partial_restore` as an unmaterialized `jax.ShapeDtypeStruct` → propagated as-is, then failing with a cryptic, low-level compile/allocation error deep in the first `train_step`.

The `scan_layers` check only validates the `scan_layers` flag, so neither case is caught for an arbitrary missing weight (a changed `emb_dim`/`mlp_dim`, a different number of layers, a dropped or renamed module, etc.).

This folds a descriptive guard into `_populate_pure_dict_from_partial`, tracking the tree path and whether the leaf sits under an NNX-only `rngs`/`dropout` subtree:

- rngs/dropout absent → filled with a deterministic default, as before (a Linen-layout checkpoint never carried it).
- any other absent-or-unmaterialized weight → governed by the new `checkpoint_missing_param_policy`:
  - `"error"` (default): raise naming the parameter path and expected shape.
  - `"warn"`: log a WARNING listing the zero-filled path and continue.

`"error"` is the default because a zero-filled or unmaterialized weight is a silentcorrectness failure. A surviving `Shape DtypeStruct` is treated the same as an absent leaf, so the previously cryptic first-`train_step` compile error becomes an explicit, path-named error at restore time.

## Tests

- `tests/unit/checkpointing_nnx_load_test.py::TestMissingParamPolicy` — present weights pass through; rng/dropout absence defaults even under `error`; a missing weight raises naming the path; a surviving `ShapeDtypeStruct` is treated as missing; `warn` zero-fills and logs; default policy is `error`.
- `tests/unit/checkpointing_nnx_missing_param_test.py` — end-to-end through the real `create_orbax_checkpoint_manager → maybe_save_checkpoint → load_state_if_possible` stack: a weight missing from the checkpoint raises naming its path, `warn` zero-fills and continues, and a matching-arch restore returns concrete arrays.

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

**Comment by xibinliu** [comment]
`_assert_no_shaped_dtype_struct` has been removed with Jacky's change. Please rebase and probably implement the check in `_populate_pure_dict_from_partial`.

**Review by NuojCheng — APPROVED** [review]

File: src/maxtext/common/checkpointing.py [hunk 1]
```diff
@@ -194,23 +194,58 @@ def _make():
...
         for k, v in abstract_pure.items()
     }
-  if partial_concrete is not None and not isinstance(partial_concrete, dict):
+  # A present, materialized leaf takes the restored value. A surviving ShapeDtypeStruct
+  # means Orbax had no matching array, so treat it as absent (handled below).
+  if partial_concrete is not None and not isinstance(partial_concrete, (dict, jax.ShapeDtypeStruct)):
     return partial_concrete
+
+  # Absent (or unmaterialized) leaf. rngs/dropout are expected to be absent from a
+  # Linen-layout checkpoint; anything else is a real weight the checkpoint should have had.
```
Comment (xibinliu) [inline src/maxtext/common/checkpointing.py]: If we handles only weights, we don't need handle rngs. Also, using hard-coded NNX_RNG_STATE_KEYS is not a safe solution. Please refer to the comments in PR [#4343](https://github.com/AI-Hypercomputer/maxtext/pull/4343) to manipulate weights/rngs with nnx.state_split

Comment (ecnal-cienet) [reply src/maxtext/common/checkpointing.py]: Done. Rebased onto #4343 and reworked to filter by Variable type instead of key names:

- Dropped `NNX_RNG_STATE_KEYS` entirely. The check now does `nnx.split_state(abstract_state, nnx.Param, ...)` and only inspects the `nnx.Param` partition — so as you said, we don't touch rngs at all. nnx.RngState/nnx.BatchStat live in #4343's nnx_aux and are never flagged.
- Kept it a minimal add-on: `_linen_items_to_nnx` and its tests are unchanged; the new `_enforce_missing_weight_policy` just runs at the restore entry points before reshaping.

A missing weight (absent, or left by Orbax as an unmaterialized `ShapeDtypeStruct`) now errors naming the parameter path, or zero-fills under `checkpoint_missing_param_policy=warn`.
